pkg/controller/endpointslicemirroring/reconciler.go

```diff
@@ -0,0 +1,306 @@
+}
+
+// reconcileByPortMapping compares the endpoints found in existing slices with
+// the list of desired endpoints and returns lists of slices to create, update,
+// and delete.
+func (r *reconciler) reconcileByPortMapping(
+	endpoints *corev1.Endpoints,
+	existingSlices []*discovery.EndpointSlice,
+	desiredSet endpointSet,
+	endpointPorts []discovery.EndpointPort,
+	addressType discovery.AddressType,
+) (slicesByAction, totalsByAction) {
+	slices := slicesByAction{}
+	totals := totalsByAction{}
+
+	// If no endpoints are desired, mark existing slices for deletion and
+	// return.
+	if desiredSet.Len() == 0 {
+		slices.toDelete = existingSlices
+		for _, epSlice := range existingSlices {
+			totals.removed += len(epSlice.Endpoints)
+		}
+		return slices, totals
+	}
+
+	if len(existingSlices) == 0 {
+		// if no existing slices, all desired endpoints will be added.
+		totals.added = desiredSet.Len()
+	} else {
+		// if >0 existing slices, mark all but 1 for deletion.
+		slices.toDelete = existingSlices[1:]
+
+		// Return early if first slice matches desired endpoints.
+		totals = totalChanges(existingSlices[0], desiredSet)
+		if totals.added == 0 && totals.updated == 0 && totals.removed == 0 {
+			return slices, totals
+		}
+	}
+
+	// generate a new slice with the desired endpoints.
+	newSlice := newEndpointSlice(endpoints, endpointPorts, addressType)
+	for desiredSet.Len() > 0 && len(newSlice.Endpoints) < int(r.maxEndpointsPerSubset) {
+		endpoint, _ := desiredSet.PopAny()
+		newSlice.Endpoints = append(newSlice.Endpoints, *endpoint)
+	}
+
+	if len(existingSlices) > 0 {
+		newSlice.Name = existingSlices[0].Name
+		slices.toUpdate = []*discovery.EndpointSlice{newSlice}
+	} else {
+		slices.toCreate = []*discovery.EndpointSlice{newSlice}
+	}
+	return slices, totals
+}
+
+// finalize creates, updates, and deletes slices as specified
+func (r *reconciler) finalize(endpoints *corev1.Endpoints, slices slicesByAction) error {
+	errs := []error{}
+
+	// If there are slices to create and delete, change the creates to updates
+	// of the slices that would otherwise be deleted.
+	for i := 0; i < len(slices.toDelete); {
+		if len(slices.toCreate) == 0 {
+			break
+		}
+		sliceToDelete := slices.toDelete[i]
+		slice := slices.toCreate[len(slices.toCreate)-1]
+		// Only update EndpointSlices that have the same AddressType as this
+		// field is considered immutable. Since Services also consider IPFamily
+		// immutable, the only case where this should matter will be the
+		// migration from IP to IPv4 and IPv6 AddressTypes, where there's a
+		// chance EndpointSlices with an IP AddressType would otherwise be
+		// updated to IPv4 or IPv6 without this check.
+		if sliceToDelete.AddressType == slice.AddressType {
+			slice.Name = sliceToDelete.Name
+			slices.toCreate = slices.toCreate[:len(slices.toCreate)-1]
+			slices.toUpdate = append(slices.toUpdate, slice)
+			slices.toDelete = append(slices.toDelete[:i], slices.toDelete[i+1:]...)
+		} else {
+			i++
```

Comment:
sorry, dumb question, but I can not understand why `i` it's only incremented here :sweat_smile:

Reply:
This is confusing, I'll work on cleaning it up. This loop is modifying the slice it's iterating through inside the loop and only iterates if it doesn't remove something from the slice. Will pull this out so it can be better tested and less confusing.

Reply:
Moved this into a separate function and added a bunch of tests - thanks for catching this rather confusing piece of code!